pkg/rpc/rpc.go

```diff
@@ -0,0 +1,166 @@
+package rpc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+)
+
+type JsonRpcRequest struct {
+	Method string          `json:"method"`
+	Params json.RawMessage `json:"params"`
+	Id     uint64          `json:"id"`
+}
+
+type JsonRpcResponse struct {
+	Result json.RawMessage `json:"result"`
+	Error  string          `json:"error"`
+	Id     uint64          `json:"id"`
+}
+
+type Error struct {
+	cause error
+}
+
+func (err *Error) Error() string {
+	return fmt.Sprintf("RPC error: %s", err.cause.Error())
+}
+
+func (err *Error) Unwrap() error {
+	return err.cause
+}
+
+type RPC struct {
+	writer    io.Closer // use encoder for writing instead
+	encoder   *json.Encoder
+	encoderMu sync.Mutex
+
+	decoder *json.Decoder
+	logger  *zap.SugaredLogger
+
+	pendingRequests map[uint64]chan JsonRpcResponse
+	lastRequestId   uint64
+	requestsMu      sync.Mutex
+
+	errChannel chan struct{} // never transports a value, only closed through setErr() to signal an occurred error
+	rpcErr     *Error        // only initialised via setErr(), if a rpc (non-skippable) error has occurred
```

Comment:
What exactly do you mean by "non-skippable"? Fatal/non-recoverable?

Reply:
Yes. Changed it to Fatal/non-recoverable.